In [2]:
import pandas as pd

### Условие 1: 1 задача
Скачать файл в уроке
Считать данные с помощью pandas
Вывести на экран первые 5 строк
Посмотреть на описание признаков и на их содержание

In [3]:
df=pd.read_csv('kc_house_data.csv', encoding='windows-1251', sep=',')
df

id             date     price  bedrooms  bathrooms  \
0      7129300520  20141013T000000  221900.0         3       1.00   
1      6414100192  20141209T000000  538000.0         3       2.25   
2      5631500400  20150225T000000  180000.0         2       1.00   
3      2487200875  20141209T000000  604000.0         4       3.00   
4      1954400510  20150218T000000  510000.0         3       2.00   
...           ...              ...       ...       ...        ...   
21608   263000018  20140521T000000  360000.0         3       2.50   
21609  6600060120  20150223T000000  400000.0         4       2.50   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21611   291310100  20150116T000000  400000.0         3       2.50   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
0             1180      5650     1.0           0     0  ...      7   
1             2570      7242     2.0           0     0  ...      7   
2              770     10000     1.0           0     0  ...      6   
3             1960      5000     1.0           0     0  ...      7   
4             1680      8080     1.0           0     0  ...      8   
...            ...       ...     ...         ...   ...  ...    ...   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
0            1180              0      1955             0    98178  47.5112   
1            2170            400      1951          1991    98125  47.7210   
2             770              0      1933             0    98028  47.7379   
3            1050            910      1965             0    98136  47.5208   
4            1680              0      1987             0    98074  47.6168   
...           ...            ...       ...           ...      ...      ...   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
0     -122.257           1340        5650  
1     -122.319           1690        7639  
2     -122.233           2720        8062  
3     -122.393           1360        5000  
4     -122.045           1800        7503  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[21613 rows x 21 columns]

In [4]:
df.head()

id             date     price  bedrooms  bathrooms  sqft_living  \
0  7129300520  20141013T000000  221900.0         3       1.00         1180   
1  6414100192  20141209T000000  538000.0         3       2.25         2570   
2  5631500400  20150225T000000  180000.0         2       1.00          770   
3  2487200875  20141209T000000  604000.0         4       3.00         1960   
4  1954400510  20150218T000000  510000.0         3       2.00         1680   

   sqft_lot  floors  waterfront  view  ...  grade  sqft_above  sqft_basement  \
0      5650     1.0           0     0  ...      7        1180              0   
1      7242     2.0           0     0  ...      7        2170            400   
2     10000     1.0           0     0  ...      6         770              0   
3      5000     1.0           0     0  ...      7        1050            910   
4      8080     1.0           0     0  ...      8        1680              0   

   yr_built  yr_renovated  zipcode      lat     long  sqft_living15  \
0      1955             0    98178  47.5112 -122.257           1340   
1      1951          1991    98125  47.7210 -122.319           1690   
2      1933             0    98028  47.7379 -122.233           2720   
3      1965             0    98136  47.5208 -122.393           1360   
4      1987             0    98074  47.6168 -122.045           1800   

   sqft_lot15  
0        5650  
1        7639  
2        8062  
3        5000  
4        7503  

[5 rows x 21 columns]

#### Условие 2: 2 задача
Проведите первичный анализ данных
Изучите типы данных
Найдите количество пропущенных ячеек в данных
Посчитайте основные статистики по всем признакам и поизучайте их
Пишите выводы

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21613 entries, 0 to 21612
Data columns (total 21 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             21613 non-null  int64  
 1   date           21613 non-null  object 
 2   price          21613 non-null  float64
 3   bedrooms       21613 non-null  int64  
 4   bathrooms      21613 non-null  float64
 5   sqft_living    21613 non-null  int64  
 6   sqft_lot       21613 non-null  int64  
 7   floors         21613 non-null  float64
 8   waterfront     21613 non-null  int64  
 9   view           21613 non-null  int64  
 10  condition      21613 non-null  int64  
 11  grade          21613 non-null  int64  
 12  sqft_above     21613 non-null  int64  
 13  sqft_basement  21613 non-null  int64  
 14  yr_built       21613 non-null  int64  
 15  yr_renovated   21613 non-null  int64  
 16  zipcode        21613 non-null  int64  
 17  lat            21613 non-null  float64
 18  long  

In [6]:
df.describe()

id         price      bedrooms     bathrooms   sqft_living  \
count  2.161300e+04  2.161300e+04  21613.000000  21613.000000  21613.000000   
mean   4.580302e+09  5.400881e+05      3.370842      2.114757   2079.899736   
std    2.876566e+09  3.671272e+05      0.930062      0.770163    918.440897   
min    1.000102e+06  7.500000e+04      0.000000      0.000000    290.000000   
25%    2.123049e+09  3.219500e+05      3.000000      1.750000   1427.000000   
50%    3.904930e+09  4.500000e+05      3.000000      2.250000   1910.000000   
75%    7.308900e+09  6.450000e+05      4.000000      2.500000   2550.000000   
max    9.900000e+09  7.700000e+06     33.000000      8.000000  13540.000000   

           sqft_lot        floors    waterfront          view     condition  \
count  2.161300e+04  21613.000000  21613.000000  21613.000000  21613.000000   
mean   1.510697e+04      1.494309      0.007542      0.234303      3.409430   
std    4.142051e+04      0.539989      0.086517      0.766318      0.650743   
min    5.200000e+02      1.000000      0.000000      0.000000      1.000000   
25%    5.040000e+03      1.000000      0.000000      0.000000      3.000000   
50%    7.618000e+03      1.500000      0.000000      0.000000      3.000000   
75%    1.068800e+04      2.000000      0.000000      0.000000      4.000000   
max    1.651359e+06      3.500000      1.000000      4.000000      5.000000   

              grade    sqft_above  sqft_basement      yr_built  yr_renovated  \
count  21613.000000  21613.000000   21613.000000  21613.000000  21613.000000   
mean       7.656873   1788.390691     291.509045   1971.005136     84.402258   
std        1.175459    828.090978     442.575043     29.373411    401.679240   
min        1.000000    290.000000       0.000000   1900.000000      0.000000   
25%        7.000000   1190.000000       0.000000   1951.000000      0.000000   
50%        7.000000   1560.000000       0.000000   1975.000000      0.000000   
75%        8.000000   2210.000000     560.000000   1997.000000      0.000000   
max       13.000000   9410.000000    4820.000000   2015.000000   2015.000000   

            zipcode           lat          long  sqft_living15     sqft_lot15  
count  21613.000000  21613.000000  21613.000000   21613.000000   21613.000000  
mean   98077.939805     47.560053   -122.213896    1986.552492   12768.455652  
std       53.505026      0.138564      0.140828     685.391304   27304.179631  
min    98001.000000     47.155900   -122.519000     399.000000     651.000000  
25%    98033.000000     47.471000   -122.328000    1490.000000    5100.000000  
50%    98065.000000     47.571800   -122.230000    1840.000000    7620.000000  
75%    98118.000000     47.678000   -122.125000    2360.000000   10083.000000  
max    98199.000000     47.777600   -121.315000    6210.000000  871200.000000

In [7]:
df['sqft_lot'].min()

520

In [8]:
df.describe(include=['object'])

date
count             21613
unique              372
top     20140623T000000
freq                142

###### Выводы.
Документ df содержит следующие типы данных: float64(5 столбцов), int64(15 стобцов), object(1 столбец). В документе нет пропущенных ячеек. Документ содержит 21613 строк × 21 столбец. 

Общее количество жилых объектов 21612. 
Средняя цена объекта: 540081 ден. ед. 
Минимальная цена объекта: 75000 ден.ед.
Максимальная цена объекта:7700000 ден.ед.
Средняя площадь объектов 15106 
Минимальная площадь: 520
Максимальная площадь: 1651359
Среднее состояние объектов (по пятибальной шкале): 3,4 
Средний возраст объектов: 47 лет 
Самые старые объекты построены в 1900 году
Самые молодые объекты построены в 2015 году
Средний класс объектов: 7.65
Минимальный класс: 1.17
Максимальный класс: 13.00
Объекты в таблице делятся на удаленность от береговой линии. 
Также есть объекты которые включают подвальные помещения. 



#### Условие 3: 3 задача
Ответьте на несколько вопросов
3.1 В каком диапазоне изменяются стоимости недвижимости?
3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?
3.3 Как много домов с разными этажами в данных?
3.4 Насколько хорошие состояния у домов в данных?
3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

##### 3.1 В каком диапазоне изменяются стоимости недвижимости?

In [9]:
df['price'].value_counts()

price
350000.0     172
450000.0     172
550000.0     159
500000.0     152
425000.0     150
            ... 
607010.0       1
1362500.0      1
298800.0       1
957500.0       1
402101.0       1
Name: count, Length: 4028, dtype: int64

In [10]:
df['price'].min()

75000.0

In [11]:
df['price'].max()

7700000.0

Стоимость изменяется в диапазоне от 75000 ден. ед до 7700000 ден.ед.  

##### 3.2 Какую долю в среднем занимают жилая площадь от всей площади по всем домам?

In [12]:
df['sqft_living'].mean()/df['sqft_lot'].mean()*100

13.76781757959227

##### 3.3 Как много домов с разными этажами в данных?

In [13]:
df['floors'].value_counts()

floors
1.0    10680
2.0     8241
1.5     1910
3.0      613
2.5      161
3.5        8
Name: count, dtype: int64

##### 3.4 Насколько хорошие состояния у домов в данных?

In [14]:
df['condition'].value_counts(normalize=True)*100

condition
3    64.919262
4    26.275852
5     7.870263
2     0.795817
1     0.138805
Name: proportion, dtype: float64

##### 3.5 Найдите года, когда построили первый дом, когда построили последний дом в данных?

In [15]:
df['yr_built'].min(),df['yr_built'].max()

(1900, 2015)

#### Условие 4: 4 задача
Ответьте на несколько вопросов

4.1 Сколько в среднем стоят дома, у которых 2 спальни?
4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?
4.3 Как много домов коснулся ремонт?
4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

#### 4.1 Сколько в среднем стоят дома, у которых 2 спальни?

In [16]:
df2=df[df['bedrooms']==2]
df2

id             date     price  bedrooms  bathrooms  \
2      5631500400  20150225T000000  180000.0         2       1.00   
11     9212900260  20140527T000000  468000.0         2       1.00   
18       16000397  20141205T000000  189000.0         2       1.00   
23     8091400200  20140516T000000  252700.0         2       1.50   
31     2426039314  20141201T000000  280000.0         2       1.50   
...           ...              ...       ...       ...        ...   
21586  2767604724  20141015T000000  505000.0         2       2.50   
21588  2767600688  20141113T000000  414500.0         2       1.50   
21595  1972201967  20141031T000000  520000.0         2       2.25   
21610  1523300141  20140623T000000  402101.0         2       0.75   
21612  1523300157  20141015T000000  325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
2              770     10000     1.0           0     0  ...      6   
11            1160      6000     1.0           0     0  ...      7   
18            1200      9850     1.0           0     0  ...      7   
23            1070      9643     1.0           0     0  ...      7   
31            1190      1265     3.0           0     0  ...      7   
...            ...       ...     ...         ...   ...  ...    ...   
21586         1430      1201     3.0           0     0  ...      8   
21588         1210      1278     2.0           0     0  ...      8   
21595         1530       981     3.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
2             770              0      1933             0    98028  47.7379   
11            860            300      1942             0    98115  47.6900   
18           1200              0      1921             0    98002  47.3089   
23           1070              0      1985             0    98030  47.3533   
31           1190              0      2005             0    98133  47.7274   
...           ...            ...       ...           ...      ...      ...   
21586        1430              0      2009             0    98107  47.6707   
21588        1020            190      2007             0    98117  47.6756   
21595        1480             50      2006             0    98103  47.6533   
21610        1020              0      2009             0    98144  47.5944   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
2     -122.233           2720        8062  
11    -122.292           1330        6000  
18    -122.210           1060        5095  
23    -122.166           1220        8386  
31    -122.357           1390        1756  
...        ...            ...         ...  
21586 -122.381           1430        1249  
21588 -122.375           1210        1118  
21595 -122.346           1530        1282  
21610 -122.299           1020        2007  
21612 -122.299           1020        1357  

[2760 rows x 21 columns]

In [17]:
mean_price=df2['price'].mean()
mean_price

401372.681884058

##### 4.2 Какая в среднем общая площадь домов, у которых стоимость больше 600 000?

In [18]:
df[df['price']>=600000]['sqft_lot'].mean()


20454.783537543197

##### 4.3 Как много домов коснулся ремонт?

In [19]:
df[df['yr_renovated']!=0].shape[0]

914

##### 4.4 Насколько в среднем стоимость домов с оценкой grade домов выше 10 отличается от стоимости домов с оценкой grade меньше 4?

In [20]:
df[df['grade']>10]['price'].mean()-df[df['grade']<4]['price'].mean()

1488885.1175298805

#### Условие 5: 5 задача
Ответьте на несколько вопросов
5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?
5.2 Выберите дом клиенту
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?
5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

##### 5.1 Выберите дом клиенту
Клиент хочет дом с видом на набережную, как минимум с тремя ванными и с подвалом. Сколько вариантов есть у клиента?


In [21]:
df_end=df[(df['waterfront']==1)&
   (df['bathrooms']>=3)&
   ~(df['sqft_basement']==0)]
df_end.shape[0]

41

##### 5.2 Выберите дом клиенту
Клиент хочет дом либо с очень красивым видом из окна, либо с видом на набережную, в очень хорошем состоянии и год постройки не меньше 1980 года. В какой ценовом диапазоне будут дома?

In [26]:
df3=df[df['yr_built']>=1980]
df3

id             date      price  bedrooms  bathrooms  \
4      1954400510  20150218T000000   510000.0         3       2.00   
5      7237550310  20140512T000000  1225000.0         4       4.50   
6      1321400060  20140627T000000   257500.0         3       2.25   
9      3793500160  20150312T000000   323000.0         3       2.50   
16     1875500060  20140731T000000   395000.0         3       2.00   
...           ...              ...        ...       ...        ...   
21608   263000018  20140521T000000   360000.0         3       2.50   
21609  6600060120  20150223T000000   400000.0         4       2.50   
21610  1523300141  20140623T000000   402101.0         2       0.75   
21611   291310100  20150116T000000   400000.0         3       2.50   
21612  1523300157  20141015T000000   325000.0         2       0.75   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
4             1680      8080     1.0           0     0  ...      8   
5             5420    101930     1.0           0     0  ...     11   
6             1715      6819     2.0           0     0  ...      7   
9             1890      6560     2.0           0     0  ...      7   
16            1890     14040     2.0           0     0  ...      7   
...            ...       ...     ...         ...   ...  ...    ...   
21608         1530      1131     3.0           0     0  ...      8   
21609         2310      5813     2.0           0     0  ...      8   
21610         1020      1350     2.0           0     0  ...      7   
21611         1600      2388     2.0           0     0  ...      8   
21612         1020      1076     2.0           0     0  ...      7   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
4            1680              0      1987             0    98074  47.6168   
5            3890           1530      2001             0    98053  47.6561   
6            1715              0      1995             0    98003  47.3097   
9            1890              0      2003             0    98038  47.3684   
16           1890              0      1994             0    98019  47.7277   
...           ...            ...       ...           ...      ...      ...   
21608        1530              0      2009             0    98103  47.6993   
21609        2310              0      2014             0    98146  47.5107   
21610        1020              0      2009             0    98144  47.5944   
21611        1600              0      2004             0    98027  47.5345   
21612        1020              0      2008             0    98144  47.5941   

          long  sqft_living15  sqft_lot15  
4     -122.045           1800        7503  
5     -122.005           4760      101930  
6     -122.327           2238        6819  
9     -122.031           2390        7570  
16    -121.962           1890       14018  
...        ...            ...         ...  
21608 -122.346           1530        1509  
21609 -122.362           1830        7200  
21610 -122.299           1020        2007  
21611 -122.069           1410        1287  
21612 -122.299           1020        1357  

[9279 rows x 21 columns]

In [28]:
df_end2=df3[(df3['view']==4)|(df3['waterfront']==1)&(df3['condition']==5)]
df_end2

id             date      price  bedrooms  bathrooms  \
246    2025069065  20140929T000000  2400000.0         4       2.50   
300    3225069065  20140624T000000  3075000.0         4       5.00   
409    1560930070  20140911T000000   840000.0         4       3.50   
757    9273200140  20150121T000000  1310000.0         2       2.25   
1164   1247600105  20141020T000000  5110800.0         5       5.25   
...           ...              ...        ...       ...        ...   
20767  8043700105  20150417T000000  2300000.0         4       4.00   
20923  7852070210  20140527T000000  1149000.0         4       3.00   
21050  2424059170  20150219T000000   900000.0         5       6.00   
21201   518500460  20141008T000000  2230000.0         3       3.50   
21576  9253900271  20150107T000000  3567000.0         5       4.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
246           3650      8354     1.0           1     4  ...      9   
300           4550     18641     1.0           1     4  ...     10   
409           2840     40139     1.0           0     4  ...     10   
757           3950      3938     2.0           0     4  ...     10   
1164          8010     45517     2.0           1     4  ...     12   
...            ...       ...     ...         ...   ...  ...    ...   
20767         4360      8175     2.5           1     4  ...     10   
20923         5940     11533     2.0           0     4  ...     11   
21050         7120     40806     2.0           0     4  ...     12   
21201         3760      5634     2.0           1     4  ...     11   
21576         4850     10584     2.0           1     4  ...     10   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
246          1830           1820      2000             0    98074  47.6338   
300          2600           1950      2002             0    98074  47.6053   
409          2840              0      1986             0    98038  47.4010   
757          2910           1040      1991             0    98116  47.5912   
1164         5990           2020      1999             0    98033  47.6767   
...           ...            ...       ...           ...      ...      ...   
20767        3940            420      2007             0    98008  47.5724   
20923        4950            990      2004             0    98065  47.5443   
21050        5480           1640      2007             0    98006  47.5451   
21201        2830            930      2014             0    98056  47.5285   
21576        3540           1310      2007             0    98008  47.5943   

          long  sqft_living15  sqft_lot15  
246   -122.072           3120       18841  
300   -122.077           4550       19508  
409   -122.026           3180       36852  
757   -122.384           3220        4500  
1164  -122.211           3430       26788  
...        ...            ...         ...  
20767 -122.104           2670        8525  
20923 -121.870           4240       12813  
21050 -122.114           3440       36859  
21201 -122.205           3560        5762  
21576 -122.110           3470       18270  

[112 rows x 21 columns]

##### В какой ценовом диапазоне будут дома?

In [41]:
df_end2['price'].min(),df_end2['price'].max()

(380000.0, 6885000.0)

##### 5.3 Выберите дом клиенту
Клиент хочет дом без подвала, с двумя этажами, стоимостью до 150000. Какая оценка по состоянию у таких домов в среднем?

In [43]:
df5=df[(df['sqft_basement']==0)&
   (df['floors']==2)&
   (df['price']<150000)
   ]
df5

id             date     price  bedrooms  bathrooms  \
2589   5061300030  20140508T000000  134000.0         2       1.50   
10105  5466310060  20150324T000000  139500.0         2       1.50   
15610  7614100080  20150211T000000  140000.0         3       1.75   
18332  9272202260  20140924T000000  130000.0         3       1.00   
18939  4239400300  20141129T000000   90000.0         3       1.00   
18988  7399100210  20141126T000000  140000.0         3       1.50   

       sqft_living  sqft_lot  floors  waterfront  view  ...  grade  \
2589           980      5000     2.0           0     0  ...      7   
10105         1230      1561     2.0           0     0  ...      7   
15610         1270      8991     2.0           0     0  ...      7   
18332         1200      7000     2.0           0     0  ...      7   
18939          980      2490     2.0           0     0  ...      6   
18988         1200      2002     2.0           0     0  ...      8   

       sqft_above  sqft_basement  yr_built  yr_renovated  zipcode      lat  \
2589          980              0      1922          2003    98014  47.7076   
10105        1230              0      1983             0    98042  47.3768   
15610        1270              0      1981             0    98042  47.3563   
18332        1200              0      1908             0    98116  47.5883   
18939         980              0      1969             0    98092  47.3170   
18988        1200              0      1966             0    98055  47.4659   

          long  sqft_living15  sqft_lot15  
2589  -121.359           1040        5000  
10105 -122.149           1660        2243  
15610 -122.149           1270        8993  
18332 -122.384           3290        6000  
18939 -122.182            980        3154  
18988 -122.189           1270        1848  

[6 rows x 21 columns]

In [44]:
df5['condition'].mean()

2.8333333333333335